src/System.Slices/src/System/Span.cs

```diff
@@ -0,0 +1,293 @@
+    public struct Span<T> : IEnumerable<T>
+    {
+        /// <summary>A managed array/string; or null for native ptrs.</summary>
+        readonly object  m_object;
+        /// <summary>An byte-offset into the array/string; or a native ptr.</summary>
+        readonly UIntPtr m_offset;
+        /// <summary>Fetches the number of elements this Span contains.</summary>
+        public readonly int Length;
+
+        /// <summary>
+        /// Creates a new span over the entirety of the target array.
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown if the 'array' parameter is null.
+        /// </exception>
+        public Span(T[] array)
+        {
+            Contract.Requires(array != null);
+            m_object = array;
+            m_offset = new UIntPtr((uint)SpanHelpers<T>.OffsetToArrayData);
+            Length = array.Length;
+        }
+
+        /// <summary>
+        /// Creates a new span over the portion of the target array beginning
+        /// at 'start' index.
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="start">The index at which to begin the span.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown if the 'array' parameter is null.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified start index is not in range (&lt;0 or &gt;&eq;length).
+        /// </exception>
+        public Span(T[] array, int start)
+        {
+            Contract.Requires(array != null);
+            Contract.RequiresInInclusiveRange(start, array.Length);
+            if (start < array.Length) {
+                m_object = array;
+                m_offset = new UIntPtr(
+                    (uint)(SpanHelpers<T>.OffsetToArrayData + (start * PtrUtils.SizeOf<T>())));
+                Length = array.Length - start;
+            }
+            else {
+                m_object = null;
+                m_offset = UIntPtr.Zero;
+                Length = 0;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new span over the portion of the target array beginning
+        /// at 'start' index and ending at 'end' index (exclusive).
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="start">The index at which to begin the span.</param>
+        /// <param name="end">The index at which to end the span (exclusive).</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown if the 'array' parameter is null.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified start or end index is not in range (&lt;0 or &gt;&eq;length).
+        /// </exception>
+        public Span(T[] array, int start, int end)
```

Comment:
:exclamation: .NET APIs representing a span are specified by start and count. The constructor needs to follow this pattern. If you also want to provide a way to construct a span from start and end indexes, you can provide a static `Span<T>.FromBounds` helper method.